### PR TITLE
Add devicetopology

### PR DESCRIPTION
**Pull request by MSxDOS**

### CODE REVIEW

File: src/um/devicetopology.rs [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+}}
+ENUM!{enum ConnectorType {
+    Unknown_Connector = 0,
+    Physical_Internal = 1,
+    Physical_External = 2,
+    Software_IO = 3,
+    Software_Fixed = 4,
+    Network = 5,
+}}
+RIDL!(#[uuid(0x28f54685, 0x06fd, 0x11d2, 0xb2, 0x7a, 0x00, 0xa0, 0xc9, 0x22, 0x31, 0x96)]
```
Comment (retep998) [inline src/um/devicetopology.rs]: Please use `{}` instead of `()` for item level macros such as `RIDL!`.

**Review by retep998 — CHANGES_REQUESTED** [review]

File: src/um/devicetopology.rs (hunk 1 above)
Comment (MSxDOS) [reply src/um/devicetopology.rs]: @🐇 Done.

File: src/um/devicetopology.rs [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+    ) -> HRESULT,
+    fn SetLevelAllChannels(
+        aLevelsDB: *mut c_float,
+        cChannels: ULONG,
+        pguidEventContext: LPCGUID,
+    ) -> HRESULT,
+}}
+RIDL!{#[uuid(0x7fb7b48f, 0x531d, 0x44a2, 0xbc, 0xb3, 0x5a, 0xd5, 0xa1, 0x34, 0xb3, 0xdc)]
+interface IAudioVolumeLevel(IAudioVolumeLevelVtbl): IPerChannelDbLevel(IPerChannelDbLevelVtbl) {}
+}
```
Comment (retep998) [inline src/um/devicetopology.rs]: This closing brace should be on the previous line.
```Rust
interface IAudioVolumeLevel(IAudioVolumeLevelVtbl): IPerChannelDbLevel(IPerChannelDbLevelVtbl) {}}
```

File: src/um/devicetopology.rs [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+}}
+RIDL!{#[uuid(0xa2b1a1d9, 0x4db3, 0x425d, 0xa2, 0xb2, 0xbd, 0x33, 0x5c, 0xb3, 0xe2, 0xe5)]
+interface IAudioBass(IAudioBassVtbl): IPerChannelDbLevel(IPerChannelDbLevelVtbl) {}
+}
+RIDL!{#[uuid(0x5e54b6d7, 0xb44b, 0x40d9, 0x9a, 0x9e, 0xe6, 0x91, 0xd9, 0xce, 0x6e, 0xdf)]
+interface IAudioMidrange(IAudioMidrangeVtbl): IPerChannelDbLevel(IPerChannelDbLevelVtbl) {}
+}
+RIDL!{#[uuid(0x0a717812, 0x694e, 0x4907, 0xb7, 0x4b, 0xba, 0xfa, 0x5c, 0xfd, 0xca, 0x7b)]
+interface IAudioTreble(IAudioTrebleVtbl): IPerChannelDbLevel(IPerChannelDbLevelVtbl) {}
+}
```
Comment (retep998) [inline src/um/devicetopology.rs]: Same for these.

File: src/um/devicetopology.rs [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+    fn GetConnectorIdConnectedTo(
+        ppwstrConnectorId: *mut LPWSTR,
+    ) -> HRESULT,
+    fn GetDeviceIdConnectedTo(
+        ppwstrDeviceId: *mut LPWSTR,
+    ) -> HRESULT,
+}}
+RIDL!{#[uuid(0x82149a85, 0xdba6, 0x4487, 0x86, 0xbb, 0xea, 0x8f, 0x7f, 0xef, 0xcc, 0x71)]
+interface ISubunit(ISubunitVtbl): IUnknown(IUnknownVtbl) {}
+}
```
Comment (retep998) [inline src/um/devicetopology.rs]: And here.

**Review by retep998 — CHANGES_REQUESTED** [review]

**Review by retep998 — APPROVED** [review]

**Pull request merged**